### PR TITLE
KVM: Improvements on upload direct download certificates

### PR DESCRIPTION
**Pull request by nvazquez**
## Description
There are some issues reported on the direct download feature for KVM:
-  The API method 'uploadTemplateDirectDownloadCertificate' always returns 'success:true' despite the ceritificate has been properly uploaded or not
- The certificate provided as parameter on 'uploadTemplateDirectDownloadCertificate' is not verified before being sent to hosts. This causes that hosts could not set up the certificates and fail the operation. 

To tackle these issues, several improvements for the KVM direct download feature have been introduced:
- Perform sanity check before sending certificate to hosts:
   - Verify the provided certificate is valid (right format) and not expired
- Report cases of failure when uploading certificates to hosts.
   - Return 'success:false' if the certificate cannot be uploaded to the hosts and log the issue properly
- Not issue-related improvements
   - Move the KVM direct download logic to KVM code
   - Add marvin tests

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
2 KVM hosts
- Upload valid certificate, verify certificate is imported on both hosts -> OK
- Upload invalid certificate -> ERROR
- Upload existing alias certificate -> ERROR
- Remove certificates
- Try direct download

### CODE REVIEW

**Comment by rhtyd** [comment]
@nvazquez please check build failure

**Comment by nvazquez** [comment]
Done thanks @rhtyd
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: agent/src/com/cloud/agent/Agent.java [hunk 1]
```diff
@@ -697,16 +697,34 @@ private Answer setupDirectDownloadCertificate(SetupDirectDownloadCertificate cmd
 
         final String keyStoreFile = agentFile.getParent() + "/" + KeyStoreUtils.KS_FILENAME;
 
-        String cerFile = agentFile.getParent() + "/" + certificateName + ".cer";
-        Script.runSimpleBashScript(String.format("echo '%s' > %s", certificate, cerFile));
+        String cerFile = agentFile.getParent() + "/" + "CSCERTIFICATE-" + certificateName;
 
+        s_logger.debug("Creating temporary certificate file into: " + cerFile);
+        int result = Script.runSimpleBashScriptForExitValue(String.format("echo '%s' > %s", certificate, cerFile));
```
Comment (DaanHoogland) [inline agent/src/com/cloud/agent/Agent.java]: handling a result is always good 👍

File: agent/src/com/cloud/agent/Agent.java [hunk 2]
```diff
@@ -697,16 +697,34 @@ private Answer setupDirectDownloadCertificate(SetupDirectDownloadCertificate cmd
 
         final String keyStoreFile = agentFile.getParent() + "/" + KeyStoreUtils.KS_FILENAME;
 
-        String cerFile = agentFile.getParent() + "/" + certificateName + ".cer";
-        Script.runSimpleBashScript(String.format("echo '%s' > %s", certificate, cerFile));
+        String cerFile = agentFile.getParent() + "/" + "CSCERTIFICATE-" + certificateName;
```
Comment (DaanHoogland) [inline agent/src/com/cloud/agent/Agent.java]: what is the file prefix for?

File: agent/src/com/cloud/agent/Agent.java [hunk 3]
```diff
@@ -697,16 +697,34 @@ private Answer setupDirectDownloadCertificate(SetupDirectDownloadCertificate cmd
...
+        int result = Script.runSimpleBashScriptForExitValue(String.format("echo '%s' > %s", certificate, cerFile));
+        if (result != 0) {
+            return new Answer(cmd, false, "Could not create the certificate file on path: " + cerFile);
+        }
+
+        s_logger.debug("Retrieving keystore password");
         String privatePasswordFormat = "sed -n '/keystore.passphrase/p' '%s' 2>/dev/null  | sed 's/keystore.passphrase=//g' 2>/dev/null";
         String privatePasswordCmd = String.format(privatePasswordFormat, agentFile.getAbsolutePath());
         String privatePassword = Script.runSimpleBashScript(privatePasswordCmd);
+        if (org.apache.commons.lang.StringUtils.isBlank(privatePassword)) {
```
Comment (DaanHoogland) [inline agent/src/com/cloud/agent/Agent.java]: we can't use an unprotected keystore on a private system?

File: agent/src/com/cloud/agent/Agent.java (hunk 2 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: This is to make sure the certificate is created into a non existing file

File: agent/src/com/cloud/agent/Agent.java (hunk 3 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Actually the check is because we're using the existing cloud.jks keystore and it is password protected

File: agent/src/com/cloud/agent/Agent.java (hunk 1 above)
Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Thanks

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: engine/schema/resources/META-INF/db/schema-41110to41120.sql [hunk 4]
```diff
@@ -27,4 +27,28 @@ INSERT IGNORE INTO `cloud`.`guest_os_hypervisor` (uuid,hypervisor_type, hypervis
...
+
+CREATE TABLE `cloud`.`direct_download_certificate_host_map` (
+  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+  `certificate_id` bigint(20) unsigned NOT NULL,
+  `host_id` bigint(20) unsigned NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `fk_direct_download_certificate_host_map__host_id` (`host_id`),
+  KEY `fk_direct_download_certificate_host_map__certificate_id` (`certificate_id`),
+  CONSTRAINT `fk_direct_download_certificate_host_map__host_id` FOREIGN KEY (`host_id`) REFERENCES `host` (`id`) ON DELETE CASCADE,
+  CONSTRAINT `fk_direct_download_certificate_host_map__certificate_id` FOREIGN KEY (`certificate_id`) REFERENCES `direct_download_certificate` (`id`) ON DELETE CASCADE
```
Comment (rhtyd) [inline engine/schema/resources/META-INF/db/schema-41110to41120.sql]: @nvazquez move this to a new 4.11.2.0-4.11.3.0 db upgrade path (for that please submit a separate pr)

**Comment by nvazquez** [comment]
Thanks @rhtyd, I moved the other improvement to separate PR #3075

**Comment by rhtyd** [comment]
Let me get back to you on this @nvazquez by tomorrow.

**Comment by nvazquez** [comment]
@rhtyd I have done it but then spread it into separate PR as requested. Please check #3075. It would need to be placed on top of this one

**Comment by nvazquez** [comment]
@rhtyd @borisstoyanov @DaanHoogland is this one ok to be merged?

**Comment by rhtyd** [comment]
@nvazquez LGTM, I'll review the other PR soon. However, I think direct download feature is specific for KVM, the code/changes should be moved to libvirtcomputing resource than in Agent/AgentShell classes. The CA related changes exist in Agent/AgentShell because those certificates secure the agent-mgmt server connection (that could be any kind of agent than just kvm agent).

**Comment by nvazquez** [comment]
@rhtyd @borisstoyanov 4.11 rebased and logic moved to KVM plugin code

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@borisstoyanov @nvazquez - is this ready for merge or do you need to perform more tests and reviews?

**Comment by rhtyd** [comment]
Closed-reopened PR to re-kick Travis.

**Comment by rhtyd** [comment]
@blueorangutan package

File: agent/src/com/cloud/agent/Agent.java [hunk 5]
```diff
@@ -685,31 +682,6 @@ protected void processRequest(final Request request, final Link link) {
         }
     }
 
-    private Answer setupDirectDownloadCertificate(SetupDirectDownloadCertificate cmd) {
```
Comment (rafaelweingartner) [inline agent/src/com/cloud/agent/Agent.java]: So, this is not used anymore?

Comment (nvazquez) [reply agent/src/com/cloud/agent/Agent.java]: Yes, but moved to specific KVM code. Let me update the description

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
Ping @rafaelweingartner could you finish the review?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRevokeDirectDownloadCertificateWrapper.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        return agentFile;
+    }
+
+    /**
+     * Get the property 'keystore.passphrase' value from agent.properties file
+     */
+    private String getKeystorePassword(File agentFile) {
+        s_logger.debug("Retrieving keystore password from agent.properties file");
+        String privatePasswordFormat = "sed -n '/keystore.passphrase/p' '%s' 2>/dev/null  | sed 's/keystore.passphrase=//g' 2>/dev/null";
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRevokeDirectDownloadCertificateWrapper.java]: Can't you use a more elegant method to retrieve these properties?

There are Java facilities that can help you load and parse key/value property file.

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRevokeDirectDownloadCertificateWrapper.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                return new Answer(command, false, "No password found for keystore: " + KeyStoreUtils.KS_FILENAME);
+            }
+
+            final String keyStoreFile = getKeyStoreFilePath(agentFile);
+
+            String checkCmd = String.format("keytool -list -alias %s -keystore %s -storepass %s",
+                    certificateAlias, keyStoreFile, privatePassword);
+            int existsCmdResult = Script.runSimpleBashScriptForExitValue(checkCmd);
+            if (existsCmdResult == 1) {
+                s_logger.debug("Certificate alias " + certificateAlias + " does not exist, no need to revoke it");
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRevokeDirectDownloadCertificateWrapper.java]: I would consider this message as an error, and not a debug one.

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtSetupDirectDownloadCertificateCommandWrapper.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        return agentFile;
+    }
+
+    /**
+     * Get the property 'keystore.passphrase' value from agent.properties file
+     */
+    private String getKeystorePassword(File agentFile) {
+        s_logger.debug("Retrieving keystore password from agent.properties file");
+        String privatePasswordFormat = "sed -n '/keystore.passphrase/p' '%s' 2>/dev/null  | sed 's/keystore.passphrase=//g' 2>/dev/null";
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtSetupDirectDownloadCertificateCommandWrapper.java]: Same suggestion as the one before.

File: server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java [hunk 9]
```diff
@@ -313,14 +322,64 @@ private DirectDownloadCommand getDirectDownloadCommandFromProtocol(DownloadProto
...
+     */
+    protected void certificateSanity(String certificatePem) {
+        Certificate certificate = getCertificateFromString(certificatePem);
+
+        if (certificate instanceof X509CertImpl) {
+            X509CertImpl x509Cert = (X509CertImpl) certificate;
+            try {
+                x509Cert.checkValidity();
+            } catch (CertificateExpiredException | CertificateNotYetValidException e) {
+                e.printStackTrace();
```
Comment (rafaelweingartner) [inline server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: A log message is better than printing the stack trace.

File: server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java [hunk 10]
```diff
@@ -313,14 +322,64 @@ private DirectDownloadCommand getDirectDownloadCommandFromProtocol(DownloadProto
...
+                throw new CloudRuntimeException("Certificate is invalid. Please provide a valid certificate. Error: " + e.getMessage());
+            }
+            if (x509Cert.getSubjectDN() != null) {
+                s_logger.info("Valid certificate for domain name: " + x509Cert.getSubjectDN().getName());
```
Comment (rafaelweingartner) [inline server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: This looks more like a debug message.

BTW, what does this message mean?
If the domain name is not null, then you call it valid? That sounds a bit odd. Could you explain that a bit further?

File: server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java [hunk 11]
```diff
@@ -313,14 +322,64 @@ private DirectDownloadCommand getDirectDownloadCommandFromProtocol(DownloadProto
...
+        certificateSanity(certificatePem);
+
+        s_logger.info("Attempting to upload certificate: " + alias + " to " + hosts.size() + " hosts");
         if (CollectionUtils.isNotEmpty(hosts)) {
             for (HostVO host : hosts) {
-                if (!uploadCertificate(certificateCer, certificateName, host.getId())) {
-                    throw new CloudRuntimeException("Uploading certificate " + certificateName + " failed on host: " + host.getId());
+                if (!uploadCertificate(certificatePem, alias, host.getId())) {
+                    s_logger.error("Could not upload certificate on host: " + host.getName() + " (" + host.getUuid() + ")");
+                    throw new CloudRuntimeException("Uploading certificate " + alias + " failed on host: " + host.getUuid());
```
Comment (rafaelweingartner) [inline server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: What if you used the same message in the log "error" method, and in the throw cloud runtime?

I think this message needs to contain, all information that are such as hostname, host uuid, and certificate alias

File: server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java [hunk 12]
```diff
@@ -331,14 +390,44 @@ public boolean uploadCertificateToHosts(String certificateCer, String certificat
...
-        final String prettified_cert = BEGIN_CERT + LINE_SEPARATOR + cert + LINE_SEPARATOR + END_CERT;
-        SetupDirectDownloadCertificate cmd = new SetupDirectDownloadCertificate(prettified_cert, certificateName);
+        SetupDirectDownloadCertificateCommand cmd = new SetupDirectDownloadCertificateCommand(certificate, certificateName);
         Answer answer = agentManager.easySend(hostId, cmd);
         if (answer == null || !answer.getResult()) {
+            String msg = "Certificate " + certificateName + " could not be added to host " + hostId;
+            if (answer != null) {
+                msg += " due to: " + answer.getDetails();
+            }
+            s_logger.info(msg);
```
Comment (rafaelweingartner) [inline server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: It is an error, right?

File: server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java [hunk 13]
```diff
@@ -331,14 +390,44 @@ public boolean uploadCertificateToHosts(String certificateCer, String certificat
...
+
+    @Override
+    public boolean revokeCertificateAlias(String certificateAlias, String hypervisor) {
+        HypervisorType hypervisorType = HypervisorType.getType(hypervisor);
+        List<HostVO> hosts = getRunningHostsToUploadCertificate(hypervisorType);
+        s_logger.info("Attempting to revoke certificate alias: " + certificateAlias + " from " + hosts.size() + " hosts");
+        if (CollectionUtils.isNotEmpty(hosts)) {
+            for (HostVO host : hosts) {
+                if (!revokeCertificateAliasFromHost(certificateAlias, host.getId())) {
+                    s_logger.error("Could not revoke certificate from host: " + host.getName() + " (" + host.getUuid() + ")");
```
Comment (rafaelweingartner) [inline server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: You can re-use the same message for both exception and log error message

File: server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java (hunk 10 above)
Comment (nvazquez) [reply server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: Actually it is valid if it succeeds the checkValidity() method above. Message is just informative, at this point in both cases (domain name null or not) the certificate is valid. Anyways I'll set it to debug level, or if its confusing then it can be removed as well

**Comment by nvazquez** [comment]
Rebased branch 4.11 and adressed last review comments, thanks @rafaelweingartner!

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@rafaelweingartner are you ok with this one after comments addressed?

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test
